tests/functional/test_configured_endpoint_url.py

```diff
@@ -0,0 +1,123 @@
+    for key, value in ini_dict.items():
+        if isinstance(value, dict):
+            section_str += f"{key} =\n"
+            for new_key, new_value in value.items():
+                section_str += f"  {new_key}={new_value}\n"
+        else:
+            section_str += f"{key}={value}\n"
+    return section_str + "\n"
+
+
+def create_cases():
+    with open(ENDPOINT_TESTDATA_FILE) as f:
+        test_suite = json.load(f)['testSuites'][0]
+
+    for test_case_data in test_suite['endpointUrlTests']:
+        yield {
+            'service': test_case_data['service'],
+            'profile': test_case_data['profile'],
+            'expected_endpoint_url': test_case_data['output'][
+                'endpointUrl'
+            ],
+            'client_args': test_suite['client_configs'].get(
+                test_case_data['client_config'], {}
+            ),
+            'config_file_contents': get_config_file_contents(
+                test_case_data['profile'], test_suite
+            ),
+            'environment': test_suite['environments'].get(
+                test_case_data['environment'], {}
+            ),
+        }
+
+
+def get_config_file_contents(profile_name, test_suite):
+    profile = test_suite['profiles'][profile_name]
+
+    profile_str = dict_to_ini_section(
+        profile,
+        section_header=f"profile {profile_name}",
+    )
+
+    services_section_name = profile.get('services', None)
+
+    if services_section_name is None:
+        return profile_str
+
+    services_section = test_suite['services'][services_section_name]
+
+    service_section_str = dict_to_ini_section(
+        services_section,
+        section_header=f'services {services_section_name}',
+    )
+
+    return profile_str + service_section_str
+
+
+def _normalize_endpoint(url):
+    split_endpoint = urlparse.urlsplit(url)
+    actual_endpoint = f"{split_endpoint.scheme}://{split_endpoint.netloc}"
+    return actual_endpoint
+
+
+class ConfiguredEndpointUrlTestCase(BaseAWSCommandParamsTest):
+    def setUp(self):
+        super(ConfiguredEndpointUrlTestCase, self).setUp()
+        self.environ["AWS_DEFAULT_REGION"] = None
+        self.files = FileCreator()
+
+    def tearDown(self):
+        super(ConfiguredEndpointUrlTestCase, self).tearDown()
+        self.files.remove_all()
+
+    def _assert_endpoint_used(
+        self, test_case
+    ):
+        self.assertEqual(
+            test_case['expected_endpoint_url'],
+            _normalize_endpoint(self.last_request_dict['url']),
+            test_case
+        )
+
+    def _create_command(self, test_case):
+        if test_case['client_args']:
+            cmd = (
+                f's3api list-buckets --profile {test_case["profile"]}'
+                f' --endpoint-url {test_case["client_args"]["endpoint_url"]}'
+            )
+        else:
+            cmd = f's3api list-buckets --profile {test_case["profile"]}'
+        return cmd
+
+    def test_resolve_configured_endpoint_url(self):
```

Comment:
Instead of using `BaseAWSCommandParamsTest`, let's try using the `CLIRunner` abstraction: https://github.com/aws/aws-cli/blob/86e86a5392f9084c2accbec18b228b15b14a5b84/tests/__init__.py#L39 It is more heavily used in the v2 codebase if you want examples on how to use it but the nice part about using it is that we will be able to parameterize these cases as it is not dependent on `unittest` and we can better guarantee we are not carrying over state between cases.